pkg/identity/identity.go

```diff
@@ -58,15 +59,16 @@ type Identity struct {
 // This structure is written as JSON to the key-value store. Do NOT modify this
 // structure in ways which are not JSON forward compatible.
 type IPIdentityPair struct {
-	IP           net.IP          `json:"IP"`
-	Mask         net.IPMask      `json:"Mask"`
-	HostIP       net.IP          `json:"HostIP"`
-	ID           NumericIdentity `json:"ID"`
-	Key          uint8           `json:"Key"`
-	Metadata     string          `json:"Metadata"`
-	K8sNamespace string          `json:"K8sNamespace,omitempty"`
-	K8sPodName   string          `json:"K8sPodName,omitempty"`
-	NamedPorts   []NamedPort     `json:"NamedPorts,omitempty"`
+	IP           net.IP           `json:"IP"`
+	Mask         net.IPMask       `json:"Mask"`
+	HostIP       net.IP           `json:"HostIP"`
+	ID           NumericIdentity  `json:"ID"`
+	Key          uint8            `json:"Key"`
+	Metadata     string           `json:"Metadata"`
+	K8sNamespace string           `json:"K8sNamespace,omitempty"`
+	K8sPodName   string           `json:"K8sPodName,omitempty"`
+	Workload     *models.Workload `json:"K8sWorkload,omitempty"`
```

Comment:
My gut feeling is that the least invasive way to do this to avoid embedding `models.Workload` and instead just to add the `Kind` directly to these structures in `pkg/identity`, `pkg/ipcache` and avoid duplicating the other workload subfields in these structures. We can use the `models.Workload` as the primary internal format for passing workload info around through function calls, but when it reaches structures like this there is no need to duplicate strings.

Reply:
For kvstore specifically, my intent here is first to avoid duplicating the information when we store it into etcd, and second to ensure backwards compatibility.

Reply:
As discussed in another comment, I will remove the namespace from Workload as it indeed duplicate information.
Regarding `workload.name`, it does not represent the pod-name but the pod controller name here. Maybe I misunderstood your message though.

Reply:
Ah you're right. It still feels a bit duplicated given that from a pod name + namespace you can derive the corresponding pod controller. I can't decide whether I think it would be better to export pod name or controller name.. An individual fault could technically affect only one pod instance, but on the other hand from a k8s deployment perspective the workload seems like the most relevant abstraction to report. We already have the Pod IP anyway so it is possible to differentiate this individual CEP vs others even if we only encoded the workload name.

Maybe I'm getting tunnel vision on this and it's OK to encode both.

Reply:
> given that from a pod name + namespace you can derive the corresponding pod controller.
That's not strictly true, and it's more nuanced than that, as it requires multiple fields (eg labels) on the pod to find the actual 'true' top-level controller. I suppose it's possible we might be able to query the correct pod from the cache using only it's name and namespace, then query the workflow however.